### PR TITLE
fix(Shard): EventEmitter listener warning

### PR DESCRIPTION
**Pull request by legendhimslef**
**Please describe the changes this PR makes and why it should be merged:**
Attaching a message listener to the parent and child during the broadcastEval and fetchClientVlaues throws a `MaxListenersExceededWarning:`
ErrorLog: https://sourceb.in/GBCHPCxLr5
The above changes were tested for 24 hours+ and it fixed the warning for all of my shards.

**Status and versioning classification:**
Code changes have been tested against the Discord API, or there are no code changes
This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

File: packages/discord.js/src/sharding/Shard.js
Comment (imranbarbhuiya) [inline packages/discord.js/src/sharding/Shard.js:420]: ig you need to document param here.

File: packages/discord.js/src/sharding/Shard.js
Comment (imranbarbhuiya) [inline packages/discord.js/src/sharding/Shard.js:424]:
```suggestion
    if (maxListeners) {
```
can't we do this?

**Review by imranbarbhuiya — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/sharding/Shard.js
Comment (legendhimslef) [reply packages/discord.js/src/sharding/Shard.js:424]: The value can be set to Infinity (or 0) to indicate an unlimited number of listeners.
source: https://nodejs.org/api/events.html#emittersetmaxlistenersn
So It's correct to check for that.

File: packages/discord.js/src/sharding/Shard.js
Comment (legendhimslef) [reply packages/discord.js/src/sharding/Shard.js:420]: These methods need to be private just like we have in the BaseClient.
source: https://discord.js.org/#/docs/main/stable/class/BaseClient
I guess we don't need to add these methods to the documentation for these classes either.
I'll leave it to maintainers.

File: packages/discord.js/src/sharding/Shard.js
Comment (imranbarbhuiya) [reply packages/discord.js/src/sharding/Shard.js:424]: Yeah I just saw it.

**Comment by imranbarbhuiya** [comment]
You need to update the typings too.

**Comment by meister03** [comment]
The error has a reason. It warns you for avoiding memory leaks, when to many listeners are opened.
There could be unwanted issues, when you eval broadcastEvals, which are never resolved. 
Besides that this is also done on collectors, maybe a `timeout` would be a fix for the upper issue. (+ your code)

File: packages/discord.js/src/sharding/Shard.js
Comment (Jiralite) [inline packages/discord.js/src/sharding/Shard.js:420]:
```suggestion
   * @param {EventEmitter|process} emitter The emitter that emits the events.
```

File: packages/discord.js/src/sharding/Shard.js
Comment (Jiralite) [inline packages/discord.js/src/sharding/Shard.js:432]:
```suggestion
   * @param {EventEmitter|process} emitter The emitter that emits the events.
```

File: packages/discord.js/typings/index.d.ts
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:1976]:
```suggestion
  private incrementMaxListeners(emitter: EventEmitter | ChildProcess): void;
```

File: packages/discord.js/typings/index.d.ts
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:1977]:
```suggestion
  private decrementMaxListeners(emitter: EventEmitter | ChildProcess): void;
```

File: packages/discord.js/typings/index.d.ts
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:2011]:
```suggestion
  private incrementMaxListeners(emitter: EventEmitter | ChildProcess): void;
```

File: packages/discord.js/typings/index.d.ts
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:2012]:
```suggestion
  private decrementMaxListeners(emitter: EventEmitter | ChildProcess): void;
```

File: packages/discord.js/src/sharding/ShardClientUtil.js
Comment (Jiralite) [inline packages/discord.js/src/sharding/ShardClientUtil.js:264]:
```suggestion
   * @param {EventEmitter|process} emitter The emitter that emits the events.
```

File: packages/discord.js/src/sharding/ShardClientUtil.js
Comment (Jiralite) [inline packages/discord.js/src/sharding/ShardClientUtil.js:252]:
```suggestion
   * @param {EventEmitter|process} emitter The emitter that emits the events.
```

**Review by Jiralite — CHANGES_REQUESTED** [review]
Please make these changes! There should be no spaces between `|` (consistency) and the tests will fail due to missing semicolons.

**Review by lucalves — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

**Review by iCrawl — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Pull request merged**
